### PR TITLE
Resolve enum literal only when dereference expression base type is enum

### PR DESCRIPTION
**Pull request by rongrong**
Test plan - unit tests

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java [hunk 1]
```diff
@@ -147,17 +147,15 @@ public static boolean isEqualComparisonExpression(Expression expression)
...
-        if (!(nodeType instanceof TypeWithName && ((TypeWithName) nodeType).getType() instanceof EnumType) || qualifiedName == null) {
-            return Optional.empty();
-        }
+
         EnumType enumType = (EnumType) ((TypeWithName) nodeType).getType();
         String enumKey = qualifiedName.getSuffix().toUpperCase(ENGLISH);
         checkArgument(enumType.getEnumMap().containsKey(enumKey), format("No key '%s' in enum '%s'", enumKey, nodeType.getDisplayName()));
         Object enumValue = enumType.getEnumMap().get(enumKey);
-        return enumValue instanceof String ? Optional.of(utf8Slice((String) enumValue)) : Optional.of(enumValue);
+        return enumValue instanceof String ? utf8Slice((String) enumValue) : enumValue;
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java:158]: This seems to be unrelated to the PR?

**Review by kaikalur — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java [hunk 2]
```diff
@@ -147,17 +147,15 @@ public static boolean isEqualComparisonExpression(Expression expression)
         return Optional.empty();
     }
 
-    public static Optional<Object> tryResolveEnumLiteral(DereferenceExpression node, Type nodeType)
+    public static Object resolveEnumLiteral(DereferenceExpression node, Type nodeType)
     {
         QualifiedName qualifiedName = DereferenceExpression.getQualifiedName(node);
-        if (!(nodeType instanceof TypeWithName && ((TypeWithName) nodeType).getType() instanceof EnumType) || qualifiedName == null) {
```
Comment (daniel-ohayon) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java:153]: looks like we're no longer checking for `qualifiedName == null`, is this fine?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 3]
```diff
@@ -302,18 +303,17 @@ public Object visitFieldReference(FieldReference node, Object context)
         @Override
         protected Object visitDereferenceExpression(DereferenceExpression node, Object context)
         {
-            Type returnType = type(node);
-            Optional<Object> maybeEnumValue = tryResolveEnumLiteral(node, returnType);
-            if (maybeEnumValue.isPresent()) {
-                return maybeEnumValue.get();
-            }
-
             Type type = type(node.getBase());
```
Comment (daniel-ohayon) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java:306]: so in the case of an enum literal `foo.bar.Color.BLUE`, both `foo.bar.Color` and `foo.bar.Color.BLUE` would be of type `ColorEnum` ?

**Review by daniel-ohayon — APPROVED** [review]
Assuming the enum unit tests are passing?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java (hunk 3 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java:306]: Yes, which makes sense right? You did set both nodes' type to the enum.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java (hunk 2 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java:153]: The check actually hide a bug in sql function where the base was rewritten from an identifier to a variable thus qualifiedName was null. If everything works correctly then qualifiedName should never be null (it's a dereference expression after all). So if we keep the check, the check should be a `checkState`.

**Review by highker — APPROVED** [review]
already reviewed in #16365
